agents_infra/agents_infra/utils/communication.py

```diff
@@ -0,0 +1,183 @@
+                api_key=api_key
+            )
+
+        if ping_url(
+                self.current_controller,
+                api_key=api_key,
+                logger=self.logger,
+                auth_token_type=self.auth_token_type
+        ):
+            return self.current_controller
+
+        remaining_urls = self._get_lower_priority_urls()
+        healthy_url = find_first_healthy_url(
+            urls=remaining_urls,
+            api_key=api_key,
+            auth_token_type=self.auth_token_type,
+            logger=self.logger
+        )
+        if healthy_url:
+            self._switch_controller(healthy_url)
+            return healthy_url
+
+        maybe_log_message(
+            'No available controller. All health checks failed.',
+            logger=self.logger
+        )
+        return None
+
+    def try_revert_primary_controller(self, api_key):
+        """
+        Attempt to revert back to the primary controller if healthy.
+
+        Returns:
+            str: Current controller URL after the check.
+        """
+        if self.current_controller == self.controller_urls[0]:
+            return self.current_controller
+
+        elapsed = get_current_time() - self.last_success_time
+        if elapsed < revert_interval:
+            return self.current_controller
+
+        higher_priority_urls = self._get_higher_priority_urls()
+        healthy_url = find_first_healthy_url(
+            higher_priority_urls,
+            api_key,
+            auth_token_type=self.auth_token_type,
+            logger=self.logger
+        )
+        if healthy_url:
+            maybe_log_message(
+                'Reverting controller: %s -> %s' % (
+                    self.current_controller,
+                    healthy_url
+                ),
+                logger=self.logger
+            )
+            self._switch_controller(healthy_url)
+
+        return self.current_controller
+
+    def _get_higher_priority_urls(self):
+        """
+        Get controllers with higher priority than the current one.
+
+        Returns:
+            list: URLs with higher priority.
+        """
+        current_index = self.controller_urls.index(self.current_controller)
+        return self.controller_urls[:current_index]
+
+    def _get_lower_priority_urls(self):
+        """
+        Get controllers with lower priority than the current one.
+
+        Returns:
+            list: URLs with lower priority.
+        """
+        current_index = self.controller_urls.index(self.current_controller)
+        return self.controller_urls[current_index + 1:]
+
+    def post_data(
+            self,
+            endpoint,
+            payload,
+            api_key=None,
+            max_retries=3,
+            delay=5,
+            timeout=5,
+            fail_silently=True,
+            **headers
+    ):
+        """
+        Post data to the active controller.
+        """
+        controller_url = self.ensure_active_controller(api_key)
```

Comment:
If `controller_url` is `None`, consider raising an exception or providing a callback, rather than just logging and returning `None`, if the calling code needs to handle this state.